x-pack/filebeat/input: Fix truncation of bodies in request tracing

chrisberkhout opened the pull request:
## Proposed commit message

```
x-pack/filebeat/input: Fix truncation of bodies in request tracing

When logging request traces, truncate the request/response body to 10%
of the maximum log file size.

Previously, bodies were truncated to the maximum file size, less 10kB.
10kB is a reasonable number for the other trace details, but space is
also required for encoding the body data as a JSON string value.

One example JSON body was 15% larger after encoding, but the 10kB
margin is 1% or less of the total limit. A body approaching the size
limit would typically generate a log entry that exceeded the limit.

Truncating large log entries to fit the file size limit means there may
only be one such entry per file. By truncating body data to 10% of the
file limit, we can expect to see entries for several request/response
pairs in each file.

The default maximum file size of 1MB gives a default maximum body size
of 100kB.

The behavior of request tracing for the HTTP Endpoint input is
unchanged: it always truncates request bodies to a size of 10kiB.
```


## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates #37826

elasticmachine commented:
Pinging @elastic/security-service-integrations (Team:Security-Service Integrations)

andrewkroh commented on CHANGELOG.next.asciidoc:
```diff
@@ -206,6 +206,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - The `_id` generation process for S3 events has been updated to incorporate the LastModified field. This enhancement ensures that the `_id` is unique. {pull}42078[42078]
 - Fix Netflow Template Sharing configuration handling. {pull}42080[42080]
 - Updated websocket retry error code list to allow more scenarios to be retried which could have been missed previously. {pull}42218[42218] 
+- Fix truncation of bodies in request tracing. {pull}42327[42327]
```
I think it would be nice to include a bit of the "how" part, like ... "by limiting bodies to 10% of the maximum file size".

chrisberkhout replied:
Yep. Done as suggested.

andrewkroh reviewed (commented):
That's an excellent commit message. Thank you.

> The default maximum file size of 1MB gives a default maximum body size
> of 100kB.

I find this behavior a bit surprising. I would have expected that if I needed to capture a full 5 MiB response body that I would need to increase the max size to something a little larger than 5 MiB, but not 10x the size. What do others think? At a minimum, I think we need to mention this behavior in the documentation associated with the tracer settings.

chrisberkhout commented:
> That's an excellent commit message. Thank you.
> 
> > The default maximum file size of 1MB gives a default maximum body size
> > of 100kB.
> 
> I find this behavior a bit surprising. I would have expected that if I needed to capture a full 5 MiB response body that I would need to increase the max size to something a little larger than 5 MiB, but not 10x the size. What do others think? At a minimum, I think we need to mention this behavior in the documentation associated with the tracer settings.

@andrewkroh I get that perspective. There are two things that make me lean towards the current version:

Often getting several responses is more important than getting full response bodies. One per file isn't a great experience. If you're getting truncated, you can double the limit (or better, shorten the page length) until you get what you need.

More importantly, since we're doing the truncation on the raw data, we need a fair bit of spare space to avoid problems. The real-world JSON I checked expanded by 15%, but it could need significantly more. A body of backslashes would double in size and bytes encoded as `\u00XX` grow by 6x.

It would be nice to truncate to a specific length as the data is written into the log but that's a much more complicated change.

I'm interested to know what others think.

If we go with this version I can improve the documentation.

kcreddy commented:
> The default maximum file size of 1MB gives a default maximum body size
of 100kB.
 
Couple of things that may impact us. 
1. With new default max body size of 100kB, we will be getting more initial diagnostics where responses are truncated and may need to go-back and request customer to increase their [resource.tracer.maxsize](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html#_resource_tracer_maxsize) if we are in need of full response (not that some APIs have the pagination links/numbers at the end of the response which are going to fall in this category.). We may need a new [resource.tracer.maxsize](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html#_resource_tracer_maxsize) default, probably 5MB to reduce this impact.
2. We don't have [resource.tracer.maxsize](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html#_resource_tracer_maxsize) exposed in most of our integrations. Until now, we have been [selectively adding them](https://github.com/elastic/integrations/pull/8290) as per the integration needs, which needs to be set globally just like `enable_request_tracer`.

chrisberkhout commented:
As discussed in the team meeting, we'll go with the current approach unless @efd6 has a better idea.

I've updated the documentation to be clear about the body limit being 10% of the file size limit. I added a `tracer.maxsize` entry in the Entity Analytics documentation.

The HTTP Endpoint [truncation](https://github.com/elastic/beats/blob/27adc68350051c939e403219c18a0bb3d6e585f8/x-pack/filebeat/input/http_endpoint/handler.go#L326-L328) always limits request bodies to 10kiB. That behavior is unchanged by this PR but I noted it in the documentation.

efd6 reviewed (commented):
The concern that I have with this is that we will see a significant growth in body sizes with this change (going from 10kB to 100kB, or 500kB, lines). I think this is likely excessive. Maybe we want that much, but I'm not convinced.

efd6 reviewed (approved):
Thanks
